diff -urp ltrace-0.5/sysdeps/linux-gnu/trace.c ltrace-0.5-pm/sysdeps/linux-gnu/trace.c
--- ltrace-0.5/sysdeps/linux-gnu/trace.c	2011-08-05 09:24:34.000000000 -0400
+++ ltrace-0.5-pm/sysdeps/linux-gnu/trace.c	2011-08-05 09:23:11.000000000 -0400
@@ -6,12 +6,19 @@
 #include <sys/types.h>
 #include "ptrace.h"
 #include <asm/unistd.h>
-#include <linux/ptrace.h>
 #include <assert.h>
 #include <sys/types.h>
 #include <sys/wait.h>
 #include <stdint.h>
 
+#if defined __ia64__
+#define ia64_fpreg ia64_fpreg_ltrace_redef
+#define pt_all_user_regs pt_all_user_regs_ltrace_redef
+#endif
+#include <linux/ptrace.h>
+#undef ia64_fpreg
+#undef pt_all_user_regs
+
 #include "ltrace.h"
 #include "options.h"
 #include "sysdep.h"
