diff -up ltrace-0.5/process_event.c\~ ltrace-0.5/process_event.c
--- ltrace-0.5/process_event.c~	2013-01-03 12:47:24.908240385 +0100
+++ ltrace-0.5/process_event.c	2013-01-03 12:53:23.422062611 +0100
@@ -122,6 +122,7 @@ static void process_clone(struct event *
 	memcpy(p, event->proc, sizeof(struct process));
 	p->pid = event->e_un.newpid;
 	p->parent = event->proc;
+	p->filename = strdup(event->proc->filename);
 
 	/* We save register values to the arch pointer, and these need
 	   to be per-thread.  */

Diff finished.  Thu Jan  3 12:54:22 2013
