diff -Nrup a/testsuite/ltrace.minor/trace-clone.c b/testsuite/ltrace.minor/trace-clone.c
--- a/testsuite/ltrace.minor/trace-clone.c	2012-12-16 18:53:45.000000000 -0700
+++ b/testsuite/ltrace.minor/trace-clone.c	2015-06-01 12:17:25.955142612 -0600
@@ -3,11 +3,14 @@
    clone called.
 
    This file was written by Yao Qi <qiyao@cn.ibm.com>.  */
+#ifndef _GNU_SOURCE
 #define _GNU_SOURCE
+#endif
 #include <stdio.h>
 #include <sys/types.h>
 #include <stdlib.h>
 #include <sched.h>
+#include <unistd.h>
 
 int child ()
 {
@@ -22,7 +25,8 @@ typedef int (* myfunc)();
 int main ()
 {
   pid_t pid;
-  static char stack[STACK_SIZE];
+  static __attribute__ ((aligned (16))) char stack[STACK_SIZE];
+
 #ifdef __ia64__
   pid = __clone2((myfunc)&child, stack, STACK_SIZE, CLONE_FS, NULL);
 #else
diff -Nrup a/testsuite/ltrace.minor/trace-irelative.exp b/testsuite/ltrace.minor/trace-irelative.exp
--- a/testsuite/ltrace.minor/trace-irelative.exp	2013-11-04 18:08:03.000000000 -0700
+++ b/testsuite/ltrace.minor/trace-irelative.exp	2015-06-01 12:30:59.737371166 -0600
@@ -54,6 +54,8 @@ set src [ltraceSource c {
 }]
 
 foreach ext {{} .pie} {
+	# AArch64 does not have ifunc support yet
+	setup_xfail aarch64*-*-*
 	set bin1 [ltraceCompile $ext $src]
 	do_tests $bin1 ""
 }
